### DISCUSSION TITLE
Exe generated by pyinstaller no longer runs after setcap cap_net_raw+eip

### DISCUSSION THREAD

**Discussion by silverv** [pyinstaller/pyinstaller#7223] [Help]
I need to set capabilities on Yukon (for it to be able to sniff the network). When I run Yukon from source then I can just setcap on the Python executable I use. In this case however I haven't found a solution. The executable file will no longer work on Ubuntu 20.04 after using
`sudo setcap cap_net_raw+eip "$(readlink -f /home/silver/Downloads/Yukon)"` on it.

How would I set this capabilty?
https://github.com/OpenCyphal-Garage/yukon/issues/79
If I remember correctly then setcap removes rpath or some path to libraries.

**Comment by rokm** [comment]
I think setting capabilities on executable causes glibc to ignore LD_LIBRARY_PATH for security reasons.

One victim of this are numpy binary extensions; but those have rpath set, so if PyInstaller preserved the shared library location layout and collected them into `numpy.libs` sub-directory instead of top-level directory, they would be discovered via rpath. To work around this, the following snippet can be used in the .spec file, inserted between `Analysis()` and `EXE()` statements:

```python
a = Analysis(...) 
pyz = PYZ(...)

# Collect numpy shared libs into numpy.libs subdirectory instead of top-level directory
def _change_numpy_libs_location(entry):
    dest, src, typecode = entry
    if typecode == 'BINARY' and dest in {'libgfortran-040039e1.so.5.0.0', 'libquadmath-96973f99.so.0.0.0', 'libopenblas64_p-r0-742d56dc.3.20.so'}:
        dest = os.path.join('numpy.libs', dest)
    return dest, src, typecode

a.binaries = (_change_numpy_libs_location(e) for e in a.binaries)

exe = EXE(...)
```

This seems to fix `numpy` error in a simple `import numpy` test program. 

But I imagine it will not suffice for more complex programs with additional binary extensions; in that case, you'll need to check each one of them and see whether its dependencies need to be moved to where rpath points (or perhaps modify the rpath), or if the rpath needs to be set, etc.

**Comment by jasonkmec** [comment]
Hello,

I have also run into this issue, and confirmed that the above workaround (adapted for the specific use-case) does indeed address the problem.

```python
def _patch_binaries(entry):
    dest, src, typecode = entry
    numpy_keys = ["libgfortran", "libquadmath", "libopenblas"]
    pyqt5_keys = ["libicudata", "libicui18n", "libicuuc", "libQt5"]
    if typecode == 'BINARY' and any(x in dest for x in numpy_keys):
        dest = os.path.join('numpy.libs', dest)
    if typecode == 'BINARY' and any(x in dest for x in pyqt5_keys):
        dest = os.path.join('PyQt5', 'Qt5', 'lib', dest)
    return dest, src, typecode

a.binaries = (_patch_binaries(e) for e in a.binaries)
```

However, I was wondering, would it be possible to include this functionality in the corresponding hooks for numpy and pyqt5? I spent some time investigating this for numpy, but it looks like these binaries are only discovered after the hook file runs, so there is no way to intercept them at the hook step. I understand that this is not a particularly common use-case, so if someone is able to point me in the right direction I would be willing to make an attempt at a more official solution.

**Reply by rokm** [reply]
Yeah, those shared libs are picked up by binary dependency analysis, not the hooks themselves. That's why they end up collected in the top-level application directory.

In the upcoming v6 release, PyInstaller will try to keep shared libraries that are bundled with wheels in their original sub-directories and symlink them to the top-level directory (instead of collecting them to top-level directory). That should alleviate the need for this work-around.

**Reply by jasonkmec** [reply]
I see, so just to confirm, the best short term solution is the above workaround?

**Reply by rokm** [reply]
Yes. 

Making symlinks to top-level directory makes no difference in the setcap case, because LD_LIBRARY_PATH (pointing to top-level directory) is not honored anyway. But do note that setcap in general will only work as long as collected binaries have rpath set (PyInstaller does not modify rpath on linux). This is usually the case with PyPI wheels, but is usually not the case if collecting from system-installed python, system-provided python packages and system-provided shared libs.

**Reply by jasonkmec** [reply]
Thank you so much for the very insightful and speedy reply!

**Comment by jasonkmec** [comment]
A slightly more elegant alternative to patching the spec file, at the cost of inflating the executable file by about 60MB (pre-compression) with redundant shared library files.

hook-numpy.py:

```python
import os
import glob
from PyInstaller.compat import EXTENSION_SUFFIXES
from PyInstaller.utils.hooks import get_package_paths

binaries = []
libs_folder = "numpy.libs"

m_dir = os.path.join(get_package_paths("numpy")[0], libs_folder)
for ext in EXTENSION_SUFFIXES:
    module_bin = glob.glob(os.path.join(m_dir, '*%s*' % ext))
    for f in module_bin:
        binaries.append((f, libs_folder))
```

hook-PyQt5.py:

```python
import os
import glob
from PyInstaller.compat import EXTENSION_SUFFIXES
from PyInstaller.utils.hooks import get_package_paths

binaries = []
libs_folder = os.path.join("PyQt5", "Qt5", "lib")

m_dir = os.path.join(get_package_paths("PyQt5")[0], libs_folder)
for ext in EXTENSION_SUFFIXES:
    module_bin = glob.glob(os.path.join(m_dir, '*%s*' % ext))
    for f in module_bin:
        binaries.append((f, libs_folder))
```

The files can be placed in a directory, which is then passed to PyInstaller using the `--additional-hooks-dir` flag (note that the file names are critical and should not be changed). I don't believe this is suitable for inclusion into the pyinstaller-hooks-contrib repo, but I'm leaving it here for anyone else who might be interested.
